### PR TITLE
chore: CoreRegistry removal from engine.network

### LINKED ISSUE #5030: Remove CoreRegistry from networking
Some MTE test flakiness is caused by race conditions in CoreRegistry access.

### PR DESCRIPTION
**Pull request by keturn**
Fixes some race conditions in MTE tests.

Fixes #5030.

### How to test

Test all the networking things!

Join server, leave server, browse server list, etc.

### CODE REVIEW

**Comment by keturn** [comment]
After asking #gestalt and seeing some [constructor injector usage](https://discord.com/channels/270264625419911192/714313228271747152/980595546366574632) with the upcoming gestalt-di, I went with that approach. It felt safer than field injection, as I know I can't create an instance without filling those values this way.

**Comment by keturn** [comment]
got some problems using this as a server; switching to Draft while I figure that out

**Comment by keturn** [comment]
okay, that was one part boring NPE and one part not knowing how to run a server

File: engine/src/main/java/org/terasology/engine/network/internal/ClientHandler.java [hunk 1]
```diff
@@ -21,16 +22,19 @@ public class ClientHandler extends ChannelInboundHandlerAdapter {
...
-    private NetworkSystemImpl networkSystem;
+    private final GameEngine gameEngine;
+    private final NetworkSystemImpl networkSystem;
+
     private ServerImpl server;
 
-    public ClientHandler(NetworkSystemImpl networkSystem) {
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    public ClientHandler(NetworkSystemImpl networkSystem, Optional<GameEngine> gameEngine) {
+        this.gameEngine = gameEngine.orElse(null);
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/network/internal/ClientHandler.java:32]: How does this behave if the engine actually happens to be `null? Will we get a reasonable error message?

**Review by skaldarnar — COMMENTED** [review]
Didn't test it yet, but code looks good to me :+1: 

Left a few questions, but none of them are really blocking.

File: engine/src/main/java/org/terasology/engine/core/modes/StateLoading.java [hunk 2]
```diff
@@ -110,6 +111,7 @@ public void init(GameEngine engine) {
         headless = context.get(DisplayDevice.class).isHeadless();
         
         CoreRegistry.setContext(context);
+        context.getValue(NetworkSystem.class).setContext(context);
```
[the remark this answers is not in the document]
Comment (keturn) [reply engine/src/main/java/org/terasology/engine/core/modes/StateLoading.java:114]: See https://github.com/MovingBlocks/Terasology/blob/6647a4e3377baf58c3f9fda4489da46b3453b5da/engine/src/main/java/org/terasology/engine/network/internal/NetworkSystemImpl.java#L778-L786.

I'm not 100% certain this addition is necessary, as I put it in during some troubleshooting and didn't confirm later. But it's near where the context is created, and it's being set in CoreRegistry which NetworkSystem _was_ using, so it's probably a good idea.

File: engine/src/main/java/org/terasology/engine/core/subsystem/common/NetworkSubsystem.java [hunk 3]
```diff
@@ -23,7 +24,7 @@ public String getName() {
 
     @Override
     public void initialise(GameEngine engine, Context rootContext) {
-        networkSystem = new NetworkSystemImpl(rootContext.get(Time.class), rootContext);
+        networkSystem = new NetworkSystemImpl((EngineTime) rootContext.get(Time.class), rootContext);
```
[the remark this answers is not in the document]
Comment (keturn) [reply engine/src/main/java/org/terasology/engine/core/subsystem/common/NetworkSubsystem.java:27]: yeah. I wondered that too.

My best guess is that `Time` is the read-only interface, and the one that gets put in Context for most common use, while `EngineTime` has other setter methods on it that the network system needs to synchronize time between client and server.

But as we can see, "please don't cast this" is not an effective form of capability control, so :shrug:

File: engine/src/main/java/org/terasology/engine/network/internal/ClientHandler.java (hunk 1 above)
Comment (keturn) [reply engine/src/main/java/org/terasology/engine/network/internal/ClientHandler.java:32]: Its only usage is here: https://github.com/MovingBlocks/Terasology/blob/6647a4e3377baf58c3f9fda4489da46b3453b5da/engine/src/main/java/org/terasology/engine/network/internal/ClientHandler.java#L37-L39

I still think it's a little suspect, but it's the nearest replacement to how `CoreRegistry.get` works.

**Review by keturn — COMMENTED** [review]
I think we'll have to ask SpotBugs to chill with the `EI_EXPOSE_REP2`

**Review by skaldarnar — APPROVED** [review]
Tested in local multiplay with a couple of bugs (which are also present on `develop`). Could connect just fine, so LGTM.

**Pull request merged**